x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/query/SparseTermsQuery.java

```diff
@@ -0,0 +1,284 @@
+        private final float termValue;
+        private final float boost;
+
+        SimpleLinearScorer(float termValue, float boost) {
+            this.termValue = termValue;
+            this.boost = boost;
+        }
+
+        @Override
+        public float score(float freq, long norm) {
+            return decodeFeatureValue(freq) * termValue * boost;
+        }
+
+        @Override
+        public Explanation explain(Explanation freq, long norm) {
+            float score = score(freq.getValue().floatValue(), 1L);
+            return Explanation.match(
+                score,
+                "score(termValue=" + decodeFeatureValue(freq.getValue().floatValue()) + ", termQueryWeight=" + termValue + ")"
+            );
+        }
+    }
+
+    // Based on Lucene package private and final class TermWeight, but with specific scorer & optimizations
+    class SparseTermWeight extends Weight {
+        private final ScoreMode scoreMode;
+        private final Similarity.SimScorer simScorer;
+        private final TermStates termStates;
+
+        SparseTermWeight(ScoreMode scoreMode, float boost, TermStates termStates) {
+            super(SparseTermsQuery.this);
+            this.scoreMode = scoreMode;
+            this.simScorer = new SimpleLinearScorer(value, boost);
+            this.termStates = termStates;
+        }
+
+        @Override
+        public Matches matches(LeafReaderContext context, int doc) throws IOException {
+            TermsEnum te = getTermsEnum(context);
+            if (te == null) {
+                return null;
+            }
+            return MatchesUtils.forField(term.field(), () -> {
+                PostingsEnum pe = te.postings(null, PostingsEnum.OFFSETS);
+                if (pe.advance(doc) != doc) {
+                    return null;
+                }
+                return new TermMatchesIterator(getQuery(), pe);
+            });
+        }
+
+        @Override
+        public String toString() {
+            return "weight(" + SparseTermsQuery.this + ")";
+        }
+
+        @Override
+        public Scorer scorer(LeafReaderContext context) throws IOException {
+            final TermsEnum termsEnum = getTermsEnum(context);
+            if (termsEnum == null) {
+                return null;
+            }
+            LeafSimScorer scorer = new LeafSimScorer(simScorer, context.reader(), term.field(), scoreMode.needsScores());
+            if (scoreMode == ScoreMode.TOP_SCORES) {
+                return new TermScorer(this, termsEnum.impacts(PostingsEnum.FREQS), scorer);
+            } else {
+                return new TermScorer(
+                    this,
+                    termsEnum.postings(null, scoreMode.needsScores() ? PostingsEnum.FREQS : PostingsEnum.NONE),
+                    scorer
+                );
+            }
+        }
+
+        @Override
+        public boolean isCacheable(LeafReaderContext ctx) {
+            return true;
+        }
```

Comment:
FeatureField isn't cacheable, it seems to me that this SHOULD be, if the segment isn't changed and the provided Term & scored value are the same, it should be cached. 

//cc @jpountz

Reply:
This is related to the other comment about `rank_features` being thought of as a way to boost scores, so it would never get cached anyway since Lucene doesn't cache scored queries. Now it would make sense to cache it so I'm happy to switch it to true, I don't think it would break anything.